orderbook/src/database/app_data.rs

```diff
@@ -0,0 +1,238 @@
+use super::*;
+use anyhow::Result;
+use futures::stream::TryStreamExt;
+use futures::StreamExt;
+use model::app_data::AppDataBlob;
+use primitive_types::{H160, H256};
+use std::borrow::Cow;
+
+#[async_trait::async_trait]
+pub trait AppDataStoring: Send + Sync {
+    async fn insert_app_data(&self, app_data: &AppDataBlob) -> Result<H256, InsertionError>;
+    fn app_data<'a>(&'a self, filter: &'a AppDataFilter) -> BoxStream<'a, Result<AppDataBlob>>;
+}
+
+/// Any default value means that this field is unfiltered.
+#[derive(Clone, Default)]
+pub struct AppDataFilter {
+    pub app_data_hash: Option<H256>,
+    pub referrer: Option<H160>,
+}
+
+#[derive(Debug)]
+pub enum InsertionError {
+    ParsingStringError(serde_json::Error),
+    DuplicatedRecord(H256),
+    AnyhowError(anyhow::Error),
+    DbError(sqlx::Error),
+}
+
+impl From<sqlx::Error> for InsertionError {
+    fn from(err: sqlx::Error) -> Self {
+        Self::DbError(err)
+    }
+}
+
+impl From<serde_json::Error> for InsertionError {
+    fn from(err: serde_json::Error) -> Self {
+        Self::ParsingStringError(err)
+    }
+}
+
+impl From<anyhow::Error> for InsertionError {
+    fn from(err: anyhow::Error) -> Self {
+        Self::AnyhowError(err)
+    }
+}
+
+#[async_trait::async_trait]
+impl AppDataStoring for Postgres {
+    async fn insert_app_data(&self, app_data_str: &AppDataBlob) -> Result<H256, InsertionError> {
+        let app_data = app_data_str.get_app_data()?;
+        const QUERY: &str = "\
+            INSERT INTO app_data (\
+                app_data_hash, app_code, referrer, file_blob)\
+                VALUES ($1, $2, $3, $4);";
+        let app_data_hash = app_data_str.sha_hash()?;
+        sqlx::query(QUERY)
+            .bind(app_data_hash.clone().as_bytes())
+            .bind(
+                app_data
+                    .app_code
+                    .unwrap_or_else(|| "".to_string())
+                    .as_bytes(),
+            )
+            .bind(
+                app_data
+                    .metadata
+                    .unwrap_or_default()
+                    .referrer
+                    .unwrap_or_default()
+                    .referrer
+                    .as_ref(),
```

Comment:
This is confusing :cold_sweat: 

I understand the model might be confusing too, but the referrer->referrer can look like a bug for someone else.
Maybe add a comment?

Also, what does `as_ref` do at the end?

triple `ref` :exploding_head:

Reply:
IF you want, we can call the referrer H160 object within the referrer object just `address` then it might be easier to read. 

In order to write it into the database as bytea, we need to hand over a slice of u8. This is what `as_ref` is doing: https://docs.rs/primitive-types/0.10.1/primitive_types/struct.H160.html

Reply:
> object just `address` then it might be easier to read.

:+1: